core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2058,6 +2058,12 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
                                 "shard term already update.  op term [" + operationPrimaryTerm + "], shardTerm [" + primaryTerm + "]";
                             primaryTerm = operationPrimaryTerm;
                             updateGlobalCheckpointOnReplica(globalCheckpoint);
+                            logger.trace(
+                                    "detected new primary with primary term [{}], resetting local checkpoint from [{}] to [{}]",
+                                    operationPrimaryTerm,
+                                    getLocalCheckpoint(),
+                                    getGlobalCheckpoint());
+                            getEngine().seqNoService().resetLocalCheckpoint(getGlobalCheckpoint());
```

Comment:
To recap a discussion we had via another channel, we do have to worry about -2 here in the case when a primary on 5.x dies and a replica on 6.x is promoted and initiates are re-sync to another 6.x replica. I pushed a d1e0ec2751d83a5c263615498929d2ae29cb1078.